rb_ws/src/buggy/scripts/auton/path_planner.py

```diff
@@ -0,0 +1,133 @@
+# WARNING: set this value to be greater than 10m/s * lookahead time (10 m/s is the upper limit
+# of NAND speed) Failure to do so can result violent u-turns in the new trajectory.
+
+FRONT_MARGIN = 30
+
+class PathPlanner():
+    def __init__(self, nominal_traj) -> None:
+        self.occupancy_grid = OccupancyGrid()
+
+        # TODO: update with NAND wheelbase
+        self.path_projector = Projector(1.3)
+
+        self.debug_passing_traj_publisher = rospy.Publisher(
+            "/auton/debug/passing_traj", NavSatFix, queue_size=1000
+        )
+
+        self.debug_splice_pt_publisher = rospy.Publisher(
+            "/auton/debug/splice_pts", NavSatFix, queue_size=1000
+        )
+
+
+        self.debug_grid_cost_publisher = rospy.Publisher(
+            "/auton/debug/grid_cost", Float64, queue_size=0
+        )
+        self.last_cmd = 0
+        self.nominal_traj = nominal_traj
+
+        # TODO: estimate this value based on the curvature of NAND's recent positions
+        self.other_steering_angle = 0
+
+    def compute_traj(
+        self,
+        other_pose: Pose, #Currently NAND's location -- To be Changed
+        other_speed: float):
+
+        # draw trajectory, such that the section of the
+        # trajectory near NAND is replace by a new segment:
+
+        # 1. the global path, at 10m behind NAND
+        # 2. NAND's projected locations, where each location is
+        # shifted to the left along the normal of the trajectory vector
+
+        other_idx = self.nominal_traj.get_closest_index_on_path(
+            other_pose.x,
+            other_pose.y)
+        #other is just NAND, for general purposes consider it other
+
+        new_segment_start_idx = self.nominal_traj.get_index_from_distance(
+                self.nominal_traj.get_distance_from_index(other_idx) - 10
+            ) #Where new path index starts, CONSTANT delta from NAND
+
+        new_segment_end_idx = self.nominal_traj.get_index_from_distance(
+                self.nominal_traj.get_distance_from_index(other_idx) + 30
+            )
+
+        # project other buggy path
+        # TODO: put other buggy command
+        other_future_poses = self.path_projector.project(
+            other_pose,
+            self.other_steering_angle,
+            other_speed,
+            LOOKAHEAD_TIME,
+            RESOLUTION)
+
+        other_future_poses_idxs = np.empty((len(other_future_poses), ))
+
+        # TODO: optimize this lookup -- how tho
+        for i in range(len(other_future_poses)):
+            other_future_poses_idxs[i] = self.nominal_traj.get_closest_index_on_path(
+                    other_future_poses[i][0],
+                    other_future_poses[i][1],
+                    start_index=other_idx)
+
+        future_pose_unit_normal = self.nominal_traj.get_unit_normal_by_index(
+            other_future_poses_idxs
+        )
+
+        # the first passing target is 10 meters backward from NAND's position
+        passing_targets = np.array(
```

Comment:
Maybe we should scale the passing target by some linear constant based on speed and just add a lower/upper bound?

Though this might be problematic if the buggy slows down or speeds up and the target point is changing even when relative to NAND.

Reply:
Only the first point in the passing targets array is 10 meters fixed distance backward, the rest are calculated from the projected position of NAND, moved to the left by a constant distance

Reply:
Right, I'm asking if we want to do the linear speed scaling for the rear point too.

Reply:
Let's make this decision after we test irl. In my opinion there's not much advantage to start the passing targets closer to NAND because the relative velocity is small. The 10 meters is really just a hyperparameter that tells the buggy how soon/late to begin the overtake. Starting the overtake early isn't going to slow you down much, if at all.

Reply:
Sounds reasonable

Reply:
Also, will remove that first point in the passing targets, since it is just the last index in the original pre-slice + 1. This will have almost no effect on the new trajectory.